### ISSUE TITLE
Cannot connect to Redshift due to hostname error

### ISSUE THREAD

**Issue by petroswork** [sqlalchemy-redshift/sqlalchemy-redshift#106]
> (psycopg2.OperationalError) server certificate for "ec2-[IP address].compute-1.amazonaws.com" does not match host name "[Hostname].redshift.amazonaws.com"

The same connection URI was fine until last Friday. I've checked that both hostnames correspond to the same IP address. If I set `'sslmode': 'prefer'` in dialect.py, the connection is successful.

According to https://docs.aws.amazon.com/redshift/latest/mgmt/connecting-ssl-support.html,
>SSL support in Amazon Redshift is strictly for encrypting the connection between your client and your cluster; it should not be relied on for authenticating the server. To authenticate the server, install the public key (.pem file) for the SSL certificate on your client and use the key to connect to your clusters.

Perhaps `sslmode: require` should be the default?

**Comment by graingert** [comment]
it seems like you're trying to connect to an ec2 instance rather than your redshift host.

**Comment by graingert** [comment]
can you paste your connection configuration

**Comment by petroswork** [comment]
Which details do you need? Also `sslmode: require` works fine.

**Comment by graingert** [comment]
The stuff you set here:
```
>>> import sqlalchemy as sa
>>> sa.create_engine('redshift+psycopg2://username@host.amazonaws.com:5439/database')
Engine(redshift+psycopg2://username@host.amazonaws.com:5439/database)
```
redact your password though.

**Comment by petroswork** [comment]
`redshift+psycopg2://petroswork:[password]@[hostname].redshift.amazonaws.com:5439/dev`

**Comment by graingert** [comment]
@petroswork I'll need the hostname

**Comment by petroswork** [comment]
I'd rather not disclose it to the public. Is it really necessary in this case?

**Comment by graingert** [comment]
You can email me direct at sqlalchemy-redshift@graingert.co.uk

**Comment by gonzalodiaz** [comment]
+1 to this issue... I have the same problem. And I can't disclosure the server URL.

**Comment by graingert** [comment]
This is looking like a configuration problem on Amazon's end... Or you're
all getting MITMd!

**Comment by gonzalodiaz** [comment]
If I try the following connection I have no problem to obtain results.:
```
self.rs_conn = psycopg2.connect(connection_string)
self.cursor = self.rs_conn.cursor()
```

To elaborate more my scenario, I'm creating an `ssh tunnel` to Redshift. Because we have an authentication server in premises. 
Note that using `psycopg2` I can connect with Redshift using the tunnel. While with sqlalchemy I'm not able to do it.

**Comment by graingert** [comment]
@gonzalodiaz you've got a different problem. (you've got a self inflicted MITM).

You should point the DNS of your redshift cluster (xyz.redshift.amazonaws.com) at your local SSH tunnel. Or you can download the certificate and manually configure your TLS settings as described by SQLalchemy

you should download the specific certificate for your redshift node, verify the domain manually then

**Comment by graingert** [comment]
Psychopg by default does not do TLS. This is why it works out of the box

**Comment by stephenpaulger** [comment]
I think I am having the same issue. I have two redshift databases that apart from running in different accounts are set up identically.

One connects fine without having to specify `connect_args={'sslmode': 'prefer'}` when calling `create_engine`. The other will fail with the message...

`sqlalchemy.exc.OperationalError: (psycopg2.OperationalError) server certificate for "IP ADDRESS" does not match host name "HOST.redshift.amazonaws.com"`

but when I specify `connect_args={'sslmode': 'prefer'}` it connects.

**Comment by graingert** [comment]
@stephenpaulger are you using an explicit IP address to connect? `{'sslmode': 'prefer'}` is not safe.

**Comment by jklukas** [comment]
Do the two clusters both have the same `sslmode` configured?

**Comment by stephenpaulger** [comment]
@graingert 
> @stephenpaulger are you using an explicit IP address to connect? {'sslmode': 'prefer'} is not safe.
The result is the same whether I am using an IP address or hostname.

@jklukas 
> Do the two clusters both have the same sslmode configured?
I assume you mean the require_ssl setting on the parameter group? In which case yes they both have require_ssl set to true.

**Comment by jrpope2014** [comment]
Hey guys, did y'all ever figure out a solution to this? We were able to switch sslmode to `sslmode=verify-ca`, but even this is not ideal, I'd REALLY like to leave it at `sslmode=verify-full`. 

I'm getting the same server certificate but it's an ip address. So `server certificate for <ip address> does not match host name "[Hostname].redshift.amazonaws.com"`. The odd part is that this ip address does not line up with the ip address of the leader node for this cluster and tcpping from the Ubuntu command line verifies that when you use `tcpping  "[Hostname].redshift.amazonaws.com" <portnumber>` the port is open, but the ip is different (as expected).

One thing that I noticed is that the logs for the cluster had this line in it just prior to when we began experiencing problems:

`Amazon Redshift cluster '<cluster_name>' has failed due to a hardware issue. The cluster is being automatically restored from the latest snapshot '<snapshot>'. Any database changes made after this time will need to be resubmitted.`

Does anyone think that it's possible the certificates for the server remained the same when it was recovered from the snapshot, so the certificate that was present was associated with the old server despite being on a new instance with a new ip? I honestly have no clue how to resolve this at this point, but it is STRONGLY preferred that we be able to maintain `sslmode=verify-full` (and it should be noted that this is what we were using before this event).

**Comment by graingert** [comment]
@jrpope2014 what do you get when you run:

```sh
openssl s_client -showcerts -connect [Hostname].redshift.amazonaws.com:5432 < /dev/null | openssl x509 -outform PEM > derp.pem
```

**Comment by jrpope2014** [comment]
@graingert I get the following:

```
140358666733216:error:140790E5:SSL routines:SSL23_WRITE:ssl handshake failure:s23_lib.c:177:
unable to load certificate
140117815387808:error:0906D06C:PEM routines:PEM_read_bio:no start line:pem_lib.c:703:Expecting: TRUSTED CERTIFICATE
```

**Comment by graingert** [comment]
@jrpope2014 how about with:

```sh
openssl s_client --no-check-certificate -showcerts -connect [Hostname].redshift.amazonaws.com:5432 < /dev/null | openssl x509 -outform PEM > derp.pem
```

**Comment by jrpope2014** [comment]
The same error with `Expecting: Trusted Certificate`, but I also get that `--no-check-certificate` is not an option. I am running openssl version `OpenSSL 1.0.1f 6 Jan 2014`

**Comment by graingert** [comment]
@jrpope2014 how about with:

```sh
openssl s_client --verify false -showcerts -connect [Hostname].redshift.amazonaws.com:5432 < /dev/null | openssl x509 -outform PEM > derp.pem
```

**Comment by jrpope2014** [comment]
```
openssl s_client -verify false -showcerts -connect [hostname].redshift.amazonaws.com:5439 < /dev/null | openssl x509 -outform PEM > derp.pem


verify depth is 0


139947504469664:error:140790E5:SSL routines:SSL23_WRITE:ssl handshake failure:s23_lib.c:177:
unable to load certificate
140456902035104:error:0906D06C:PEM routines:PEM_read_bio:no start line:pem_lib.c:703:Expecting: TRUSTED CERTIFICATE
```

(We use port 5439 for connections)

**Comment by graingert** [comment]
@jrpope2014 sounds like it's time to call AWS support.

**Comment by jrpope2014** [comment]
@graingert I agree! Haha I really appreciate the help though. I actually have a ticket open with them. They requested that I try something, I tried it, told them that it didn't work, and now I've been waiting 5 days for another response. Lol maybe they don't know either? I don't know. But it's really disconcerting. Broken SSL is not okay :(.

**Comment by jrpope2014** [comment]
@graingert to try and verify again, I did just rerun this command on a cluster that I AM able to connect successfully with `sslmode=verify-full` and received the same error.

**Comment by graingert** [comment]
@jrpope2014  ah woops, of course that's because PG doesn't handshake straight away, it waits for STARTTLS. You need something like: https://github.com/thusoy/postgres-mitm/blob/master/postgres_get_server_cert.py

**Comment by jrpope2014** [comment]
Okay, awesome, @graingert , I did in fact get a cert when using this script (had to update python to > 2.7.9 for anyone else reading this that wants to use that script).

**Comment by graingert** [comment]
@jrpope2014 can you post the cert?

**Comment by jrpope2014** [comment]
Hey @graingert apologies for the inconvenience, but I cannot post the cert because the hostname can be extracted from it but I can tell you that when running `openssl x509 -noout -in derp.pem -subject` I receive the following output:

`subject= /CN=<mystery_ip>/OU=RDS/O=Amazon.com/L=Seattle/ST=Washington/C=US`

Where the mystery ip is the ip address that kept popping up in the error message during the hostname validation step. This makes sense and is what I expected (but this is verification instead of just "going with my gut", which I really appreciate!). I also ran this against some of the databases that I AM able to use `sslmode=verify-full` with and noted that `/CN=<database_endpoint instead of any ip address>`. Are there any other commands that you would recommend I run against the cert to glean information? 

Thanks again for all the assistance here. I don't know about everyone else's issues, but I feel like this is not likely a bug in the `sqlalchemy-redshift` code, but rather something wrong in Amazon's automation for database recovery. What do you think?

**Comment by SeanTurner026** [comment]
So I'm having a similar issue. I'm trying to ssh port forward from a circle CI node with the following command (I white list the CI node's IP address with boto and configured ingress, that all works).

`ssh -o StrictHostKeyChecking=no -N -L 127.0.0.1:5439:$REDSHIFT_DNS:5439 -i $BASTION_CERT $USER@$BASTION_DNS`

And then to test that the connection is working I'm running a simple script (from the CI node) after that iterates through table names, prints the first one, and exits

```
from sqlalchemy import create_engine, inspect

engine = create_engine('redshift+psycopg2://USER:PASS@localhost:5439/DB_NAME')
inspector = inspect(engine)
for table_name in inspector.get_table_names():
    print(table_name)
    break
```

I was getting the following error

`sqlalchemy.exc.OperationalError: (psycopg2.OperationalError) server common name "*.xyz123.region.redshift.amazonaws.com" does not match host name "localhost"`

Then I added the `$REDSHIFT_DNS` to the `/etc/hosts` on my bastion (per @graingert 's advice above) 
```
$ cat /etc/hosts
127.0.0.1   localhost <REDSHIFT_DNS>
```

and I'm now receiving the following error:

`sqlalchemy.exc.OperationalError: (psycopg2.OperationalError) server closed the connection unexpectedly
	This probably means the server terminated abnormally
	before or while processing the request.`

Have I pointed the the DNS of the redshift cluster (xyz.redshift.amazonaws.com) at the local SSH tunnel using /etc/hosts correctly? Any advice would be greatly greatly appreciated 🙏
